flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultAllocatedSlotPoolTest.java

```diff
@@ -0,0 +1,315 @@
+		slots.add(otherSlot);
+
+		slotPool.addSlots(slots, 0);
+
+		slotPool.removeSlots(owner);
+
+		assertSlotPoolContainsSlots(slotPool, Collections.singleton(otherSlot));
+	}
+
+	@Test
+	public void testContainsSlots() {
+		final DefaultAllocatedSlotPool slotPool = new DefaultAllocatedSlotPool();
+		final ResourceID owner = ResourceID.generate();
+		final AllocatedSlot allocatedSlot = createAllocatedSlot(owner);
+
+		slotPool.addSlots(Collections.singleton(allocatedSlot), 0);
+
+		assertTrue(slotPool.containsSlots(owner));
+		assertFalse(slotPool.containsSlots(ResourceID.generate()));
+	}
+
+	@Test
+	public void testContainsSlot() {
+		final DefaultAllocatedSlotPool slotPool = new DefaultAllocatedSlotPool();
+		final AllocatedSlot allocatedSlot = createAllocatedSlot(null);
+
+		slotPool.addSlots(Collections.singleton(allocatedSlot), 0);
+
+		assertTrue(slotPool.containsSlot(allocatedSlot.getAllocationId()));
+		assertFalse(slotPool.containsSlot(new AllocationID()));
+	}
+
+	@Test
+	public void testReserveFreeSlot() {
+		final DefaultAllocatedSlotPool slotPool = new DefaultAllocatedSlotPool();
+		final Collection<AllocatedSlot> allSlots = createAllocatedSlots();
+		final Collection<AllocatedSlot> freeSlots = new ArrayList<>(allSlots);
+		final Iterator<AllocatedSlot> iterator = freeSlots.iterator();
+		final AllocatedSlot allocatedSlot = iterator.next();
+		iterator.remove();
+
+		slotPool.addSlots(allSlots, 0);
+
+		assertThat(slotPool.reserveFreeSlot(allocatedSlot.getAllocationId()), sameInstance(allocatedSlot));
+
+		assertSlotPoolContainsFreeSlots(slotPool, freeSlots);
+		assertSlotPoolContainsSlots(slotPool, allSlots);
+	}
+
+	@Test(expected = IllegalStateException.class)
+	public void testReserveNonFreeSlotFails() {
+		final DefaultAllocatedSlotPool slotPool = new DefaultAllocatedSlotPool();
+		final AllocatedSlot slot = createAllocatedSlot(null);
+
+		slotPool.addSlots(Collections.singleton(slot), 0);
+
+		slotPool.reserveFreeSlot(slot.getAllocationId());
+		slotPool.reserveFreeSlot(slot.getAllocationId());
+	}
+
+	@Test
+	public void testFreeReservedSlot() {
+		final DefaultAllocatedSlotPool slotPool = new DefaultAllocatedSlotPool();
+		final Collection<AllocatedSlot> slots = createAllocatedSlots();
+
+		final int initialTime = 0;
+		slotPool.addSlots(slots, initialTime);
+
+		final AllocatedSlot slot = slots.iterator().next();
+
+		slotPool.reserveFreeSlot(slot.getAllocationId());
+
+		final int releaseTime = 1;
+		assertTrue(slotPool.freeReservedSlot(slot.getAllocationId(), releaseTime).isPresent());
+		assertSlotPoolContainsFreeSlots(slotPool, slots);
+
+		for (AllocatedSlotPool.FreeSlotInfo freeSlotInfo : slotPool.getFreeSlotsInformation()) {
+			final long time;
+			if (freeSlotInfo.getAllocationId().equals(slot.getAllocationId())) {
+				time = releaseTime;
+			} else {
+				time = initialTime;
+			}
+
+			assertThat(freeSlotInfo.getFreeSince(), is(time));
+		}
+	}
+
+	@Test
+	public void testFreeFreeSlotIsIgnored() {
```

Comment:
```suggestion
	public void testFreeReservedSlotIsIgnored() {
```

Reply:
This test checks that freeing a free slot does not cause any side-effects, so the proposed name would be wrong. The method name is indeed a bit awkward; an alternative might be `testFreeingFreeSlotIsIgnored`.

Reply:
Ah ok, then let's call `testFreeingFreeSlotIsIgnored`.